### PR TITLE
add markdown files for agentic models

### PR DESCRIPTION
**Pull request by Adel-Moumen**
This PR adds markdown instructions for agentic models (e.g. Claude Code, Codex, etc.) to help them navigate the codebase.

These files are meant to evolve over time and gradually reflect the common issues LLMs may encounter when working with the SpeechBrain codebase. This first PR is intended as a prototype in that direction.

### CODE REVIEW

**Comment by Adel-Moumen** [comment]
ps: I am not a prompt engineer, and therefore I believe that things could be improved (e.g. should it be more concise etc?) but I do believe that the only way to know is to start from somewhere and then slowly update/build on top.

File: AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+**DynamicBatchSampler** (`speechbrain.dataio.sampler`) groups utterances into length-bucketed batches with a target total duration instead of a fixed batch size. Requires a `length_func` pointing to the manifest duration field. Do not pass `batch_size` when using `batch_sampler`.
+
+**CategoricalEncoder** (`speechbrain.dataio.encoder`) maps string labels to integer indices for classification. Fit with `encoder.update_from_didataset(train_data, "spk_id")`, use `encoder.encode_label_torch(label)` inside a pipeline, sanity-check with `encoder.expect_len(num_classes)`.
+
+Every recipe wires this together in a `dataio_prep(hparams)` function — follow this pattern for new recipes.
+
+## Recipe conventions
+
+Every recipe lives at `recipes/{dataset}/{task}/{mdeol}` and follows this structure:
```
Comment (pplantinga) [inline AGENTS.md:190]:
```suggestion
Every recipe lives at `recipes/{dataset}/{task}/{model}` and follows this structure:
```

File: AGENTS.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+pytest tests/unittests/
+
+# Run a specific integration test
+pytest tests/integration/ASR_CTC/ -xvs
+
+# Run all integration tests
+pytest tests/integration/ -x
+```
+
+Pre-commit hooks are configured in `.pre-commit-config.yaml` and enforce formatting/linting automatically. Always run `pre-commit run -a` before opening a PR.
```
Comment (pplantinga) [inline AGENTS.md:225]: Not sure how installing the pre-commit hooks interacts with agents here. Looks like this assumes the agent will always manually run the tests rather than installing the hook. Just wanted to check that this is what we want to do, as it seems there is some risk of the agent forgetting this part (which is not the end of the world but could be annoying I guess).

File: AGENTS.md [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+- **modules vs hparams**: objects listed under `modules:` in the YAML are registered as `nn.Module`s on the Brain (moved to device, included in DDP, saved in checkpoints). Objects accessed via `self.hparams.*` are not. Putting a trainable module only in hparams means it won't be on the right device or saved properly.
+- **Stage-dependent logic**: always check `stage` before computing validation-only metrics or applying train-only augmentation. Forgetting this causes training-time metric computation (slow) or test-time augmentation (wrong results).
+- **Batch format**: batch objects from the dataio pipeline are `PaddedBatch` instances. Access signals as `batch.sig` which returns `(tensor, lengths)` tuples. Do not index batch like a plain dict.
+- **Checkpointing**: Brain's checkpointer saves/loads modules, optimizers, schedulers, and epoch counters. If you add a new trainable module, register it with the checkpointer or it won't be saved/restored.
+- **Soundfile vs torchaudio**: SpeechBrain is migrating audio I/O from torchaudio to soundfile. Use `speechbrain.dataio.dataio.read_audio` for reading audio, not raw torchaudio calls.
```
Comment (pplantinga) [inline AGENTS.md:250]:
```suggestion
- **Soundfile vs torchaudio**: SpeechBrain uses a soundfile backend for audio I/O. Use `speechbrain.dataio.dataio.read_audio` for reading audio, not raw torchaudio calls.
```

File: AGENTS.md [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+```
+
+The `from_hparams` method downloads the model and YAML from HuggingFace, loads via HyperPyYAML, and returns a ready-to-use object.
+
+## Key dependencies
+
+- PyTorch (core)
+- HyperPyYAML (`hyperpyyaml` package — SpeechBrain's extended YAML, separate repo at `speechbrain/HyperPyYAML`)
+- soundfile (audio I/O)
+- torchaudio (some legacy audio I/O, being phased out)
```
Comment (pplantinga) [inline AGENTS.md:269]:
```suggestion
- torchaudio (basic feature transforms, resampling, etc.)
```

File: AGENTS.md [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+**PaddedBatch** (`speechbrain.dataio.batch`) is the collate function. It pads variable-length tensors and returns `PaddedData(data, lengths)` namedtuples. Always unpack: `wavs, wav_lens = batch.sig`. The `lengths` are relative (0.0–1.0). Use `SaveableDataLoader` instead of raw `DataLoader` — it supports checkpoint-resumable iteration.
+
+**DynamicBatchSampler** (`speechbrain.dataio.sampler`) groups utterances into length-bucketed batches with a target total duration instead of a fixed batch size. Requires a `length_func` pointing to the manifest duration field. Do not pass `batch_size` when using `batch_sampler`.
+
+**CategoricalEncoder** (`speechbrain.dataio.encoder`) maps string labels to integer indices for classification. Fit with `encoder.update_from_didataset(train_data, "spk_id")`, use `encoder.encode_label_torch(label)` inside a pipeline, sanity-check with `encoder.expect_len(num_classes)`.
+
+Every recipe wires this together in a `dataio_prep(hparams)` function — follow this pattern for new recipes.
+
+## Recipe conventions
```
Comment (pplantinga) [inline AGENTS.md:188]: Should we actually have additional `AGENTS.md` files in key top-level folders as well? Like one for recipes, one for tests, one for speechbrain folder itself? Just wondering if it might be helpful to have more specific instructions depending on what the agent is trying to do.

File: AGENTS.md [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+- **Minimize dependencies**: core library should stay lightweight. Recipe-specific deps go in `extra_requirements.txt`. If adding to core but the dependency is heavy, put it in `speechbrain/integrations/`.
+- **Efficiency**: operate on batches, prefer tensor ops over Python loops, avoid CPU/GPU syncs.
+
+## Common pitfalls
+
+- **HyperPyYAML is not plain YAML**: do not treat `.yaml` files as simple config. `!new:` instantiates objects, `!ref` resolves references. Editing these files requires understanding the tag system. If you break a `!ref` chain, training will crash at load time.
+- **Relative lengths, not absolute**: SpeechBrain passes relative lengths (0 to 1) for masking/padding. Do not pass absolute sample counts where relative lengths are expected.
+- **modules vs hparams**: objects listed under `modules:` in the YAML are registered as `nn.Module`s on the Brain (moved to device, included in DDP, saved in checkpoints). Objects accessed via `self.hparams.*` are not. Putting a trainable module only in hparams means it won't be on the right device or saved properly.
```
Comment (pplantinga) [inline AGENTS.md:246]:
```suggestion
- **modules vs hparams**: objects listed under `modules:` in the YAML are registered as `nn.Module`s on the Brain (moved to device, included in DDP). Objects accessed via `self.hparams.*` are not. Putting a trainable module only in hparams means it won't be on the right device or saved properly.
```

The "saved in checkpoints" is unrelated to "modules" I think.

**Review by pplantinga — COMMENTED** [review]
Looks like a good start, my main comment is that we may want to start thinking of specific agentic workflows and design special files for this, such as "adding a new feature to speechbrain core" which could be covered by an `AGENTS.md` file in the `speechbrain/` folder with more detailed instructions on how to run tests and write unittests and ensure everything is working. Or a "how to write a new recipe" file for the recipes folder, etc. Not sure if it is necessary now but could be nice to do while we are thinking about it.

**Review by pplantinga — APPROVED** [review]
LGTM!
